cmd/background-heal-ops.go

```diff
@@ -0,0 +1,141 @@
+/*
+ * Minio Cloud Storage, (C) 2019 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"time"
+
+	"github.com/minio/minio/cmd/logger"
+	"github.com/minio/minio/pkg/madmin"
+)
+
+// healTask represents what to heal along with options
+//   path: '/' =>  Heal disk formats along with metadata
+//   path: 'bucket/' or '/bucket/' => Heal bucket
+//   path: 'bucket/object' => Heal object
+type healTask struct {
+	path string
+	opts madmin.HealOpts
+	// Healing response will be sent here
+	responseCh chan healResult
+}
+
+// healResult represents a healing result with a possible error
+type healResult struct {
+	result madmin.HealResultItem
+	err    error
+}
+
+type healRoutine struct {
+	tasks chan healTask
+}
+
+func (h *healRoutine) queueHealTask(task healTask) {
+	h.tasks <- task
+}
+
+func (h *healRoutine) run() {
```

Comment:
We should add sleep before starting run() and also make it binomial to avoid starting this thread simultaneously on all servers at the same time.

Reply:
> We should add sleep before starting run() and also make it binomial to avoid starting this thread simultaneously on all servers at the same time.

Basically, we use a lock, so each healing is done by a different server but not in parallel